crates/sui-graphql-rpc/src/server/watermark_task.rs

```diff
@@ -155,36 +164,75 @@ impl Watermark {
     pub(crate) async fn new(lock: WatermarkLock) -> Self {
         let w = lock.read().await;
         Self {
-            checkpoint: w.checkpoint,
-            checkpoint_timestamp_ms: w.checkpoint_timestamp_ms,
+            hi_cp: w.hi_cp,
+            hi_cp_timestamp_ms: w.hi_cp_timestamp_ms,
             epoch: w.epoch,
+            lo_cp: w.lo_cp,
+            lo_tx: w.lo_tx,
         }
     }
 
+    #[allow(clippy::type_complexity)]
     pub(crate) async fn query(db: &Db) -> Result<Option<Watermark>, Error> {
         use checkpoints::dsl;
-        let Some((checkpoint, checkpoint_timestamp_ms, epoch)): Option<(i64, i64, i64)> = db
+        let Some(result): Option<Vec<(i64, i64, i64, Option<i64>)>> = db
             .execute(move |conn| {
                 async {
-                    conn.first(move || {
-                        dsl::checkpoints
-                            .select((dsl::sequence_number, dsl::timestamp_ms, dsl::epoch))
+                    conn.results(move || {
+                        let min_cp = dsl::checkpoints
+                            .select((
+                                dsl::sequence_number,
+                                dsl::timestamp_ms,
+                                dsl::epoch,
+                                dsl::min_tx_sequence_number,
+                            ))
+                            .order_by(dsl::sequence_number.asc())
+                            .limit(1);
+
+                        let max_cp = dsl::checkpoints
+                            .select((
+                                dsl::sequence_number,
+                                dsl::timestamp_ms,
+                                dsl::epoch,
+                                dsl::min_tx_sequence_number,
+                            ))
                             .order_by(dsl::sequence_number.desc())
+                            .limit(1);
+
+                        // Order by sequence_number, which is in the 1st position
+                        min_cp.union_all(max_cp).positional_order_by(1)
                     })
                     .await
                     .optional()
                 }
                 .scope_boxed()
             })
             .await
-            .map_err(|e| Error::Internal(format!("Failed to fetch checkpoint: {e}")))?
+            .map_err(|e| Error::Internal(format!("Failed to fetch watermark data: {e}")))?
         else {
+            // An empty response from the db is valid when indexer has not written any checkpoints
+            // to the db yet.
             return Ok(None);
         };
+
+        if result.len() != 2 {
+            return Err(Error::Internal(
+                "Expected 2 rows from watermark query".to_string(),
+            ));
+        }
```

Comment:
This seems unnecessary -- we can structure the code to not worry about there being two distinct results (see comment below).